src/Control/HTTP.php

```diff
@@ -376,156 +413,132 @@ public static function register_etag($etag)
      */
     public static function add_cache_headers($body = null)
     {
-        $cacheAge = self::$cache_age;
-
         // Validate argument
         if ($body && !($body instanceof HTTPResponse)) {
             user_error("HTTP::add_cache_headers() must be passed an HTTPResponse object", E_USER_WARNING);
             $body = null;
         }
 
-        // Development sites have frequently changing templates; this can get stuffed up by the code
-        // below.
-        if (Director::isDev()) {
-            $cacheAge = 0;
-        }
-
         // The headers have been sent and we don't have an HTTPResponse object to attach things to; no point in
         // us trying.
         if (headers_sent() && !$body) {
             return;
         }
 
-        // Populate $responseHeaders with all the headers that we want to build
-        $responseHeaders = array();
+        // Warn if already assigned cache-control headers
+        if ($body && $body->getHeader('Cache-Control')) {
+            trigger_error(
+                'Cache-Control header has already been set. '
+                . 'Please use HTTPCacheControlMiddleware API to set caching options instead.',
+                E_USER_WARNING
+            );
+            return;
+        }
 
-        $cacheControlHeaders = HTTP::config()->uninherited('cache_control');
+        $config = Config::forClass(__CLASS__);
 
+        // Get current cache control state
+        $cacheControl = HTTPCacheControlMiddleware::singleton();
 
-        // currently using a config setting to cancel this, seems to be so that the CMS caches ajax requests
-        if (function_exists('apache_request_headers') && static::config()->uninherited('cache_ajax_requests')) {
-            $requestHeaders = array_change_key_case(apache_request_headers(), CASE_LOWER);
-            if (isset($requestHeaders['x-requested-with'])
-                && $requestHeaders['x-requested-with']=='XMLHttpRequest'
-            ) {
-                $cacheAge = 0;
-            }
+        // if http caching is disabled by config, disable it - used on dev environments due to frequently changing
+        // templates and other data. will be overridden by forced publicCache() or privateCache() calls
+        if ($config->get('disable_http_cache')) {
```

Comment:
Shift this to HTTPCacheControlMiddleware

Reply:
Agreed, its deprecated but if it only applies on a code path we're no longer using in core (by avoiding `add_cache_headers()`) then it needs to be moved.